components/codec/src/number.rs

```diff
@@ -0,0 +1,2064 @@
+        write!(self, v, 8, encode_u64_le)
+    }
+
+    /// Writes a signed 64 bit integer `v` in little endian,
+    /// which is not memory-comparable.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < 8.
+    #[inline]
+    fn write_i64_le(&mut self, v: i64) -> Result<()> {
+        write!(self, v, 8, encode_i64_le)
+    }
+
+    /// Writes a 64 bit float number `v` in little endian,
+    /// which is not memory-comparable.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < 8.
+    #[inline]
+    fn write_f64_le(&mut self, v: f64) -> Result<()> {
+        write!(self, v, 8, encode_f64_le)
+    }
+
+    /// Writes an unsigned 64 bit integer `v` in VarInt encoding,
+    /// which is not memory-comparable. Returns thenumber of bytes that encoded.
+    ///
+    /// Note:
+    /// - VarInt encoding is slow, try avoid using it.
+    /// - The buffer must reserve 10 bytes for writing, although actual written bytes may be less.
+    /// - The buffer will be advanced by actual written bytes.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < 10.
+    #[inline]
+    fn write_var_u64(&mut self, v: u64) -> Result<usize> {
+        let encoded_bytes = {
+            let buf = unsafe { self.bytes_mut(MAX_VARINT64_LENGTH) };
+            if buf.len() < MAX_VARINT64_LENGTH {
+                return Err(Error::BufferTooSmall);
+            }
+            NumberCodec::encode_var_u64(buf, v)
+        };
+        unsafe { self.advance_mut(encoded_bytes) };
+        Ok(encoded_bytes)
+    }
+
+    /// Writes a signed 64 bit integer `v` in VarInt encoding,
+    /// which is not memory-comparable. Returns the number of bytes that encoded.
+    ///
+    /// Note:
+    /// - VarInt encoding is slow, try avoid using it.
+    /// - The buffer must reserve 10 bytes for writing, although actual written bytes may be less.
+    /// - The buffer will be advanced by actual written bytes.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < 10.
+    #[inline]
+    fn write_var_i64(&mut self, v: i64) -> Result<usize> {
+        let encoded_bytes = {
+            let buf = unsafe { self.bytes_mut(MAX_VARINT64_LENGTH) };
+            if buf.len() < MAX_VARINT64_LENGTH {
+                return Err(Error::BufferTooSmall);
+            }
+            NumberCodec::encode_var_i64(buf, v)
+        };
+        unsafe { self.advance_mut(encoded_bytes) };
+        Ok(encoded_bytes)
+    }
+}
+
+/// Any types who implemented `BufferWriter` also implements `BufferNumberEncoder`.
+impl<T: BufferWriter> BufferNumberEncoder for T {}
+
+#[cfg(test)]
+mod tests {
+    use protobuf::CodedOutputStream;
+    use rand;
+
+    fn get_u16_samples() -> Vec<u16> {
+        vec![
+            (::std::i16::MIN as u16),
+            (::std::i16::MIN as u16).overflowing_add(1).0,
```

Comment:
```suggestion
            (::std::i16::MIN as u16).wrapping_add(1),
```

Same for all `overflowing_xxx` below.